Improve option setting/parsing and skip

sandfox opened the pull request:
This PR improves option parsing by explicitly detting the boolean and it also modifies the behaviour for `raw` mode by skipping parsing on the BSON record.

let me know if you want this split out into 2 seperate PRs

I also couldn't decide if the "raw mode skip parsing" should sit behind a further option flag?

In terms of unscientific and useless benchmarks, average parsing speed of a 634220 record, 3.3GB dataset on my SSD 2013 macbook went from around 110 seconds to 6

timkuijsten commented:
> In terms of unscientific and useless benchmarks, average parsing speed of a 634220 record, 3.3GB dataset on my SSD 2013 macbook went from around 110 seconds to 6

This is purely in the new "raw-no-parse" mode I suppose, not because of the option parsing is it?

Why would you want explicit typing of the options instead of their standard truthfulness?

timkuijsten commented:
> I also couldn't decide if the "raw mode skip parsing" should sit behind a further option flag?

+1 for a `validateRaw` option, that defaults to false.

sandfox commented:
Yep, the super fast was in "raw-no-parse" mode, so basically just splitting the stream up into discrete BSON documents/buffers

My current use cases for skipping parsing was for counting records and record size, redistributing records to worker processes for more heavyweight processing, and other stuff where it was useful to be able to work with just the raw buffers without paying the cost of parsing at the same time.

I've added the `validateRaw` option but couldn't make the description fit on one line so indented the 2nd for clarity.

```
* validateRaw {Boolean, defaults false} only used when raw = true, whether to 
*   validate the raw Buffer before emitting by parsing it.
```

I guess the explicit typing on the options was a mix of habit, ease of reading for me and more than happy to revert it back again if you want.

I've fixed up the previous formatting issue you raised, let me know if there are any others.

timkuijsten commented:
awesome! If you can add one test for the new validateRaw option (just test an invalid BSON doc with only a length field that fails if validateRaw is false and passes if validateRaw is true), I will merge it.
